File: Pesquisa e Ordenação/aula_04_bolha_seleção_inserção/ordenacao.py
class Ordenacao:
    @staticmethod
    def bolha(lista):
        houve_troca = True
        qtd_comparacoes = 0
        qtd_trocas = 0
        while (houve_troca):    
            houve_troca = False
            for i in range (len(lista) - 1):
                qtd_comparacoes+=1
                if (lista[i] > lista[i+1]):
                    qtd_trocas+=1
                    houve_troca = True
                    tmp = lista[i]
                    lista[i] = lista[i+1]
                    lista[i+1] = tmp
                    
        return qtd_comparacoes, qtd_trocas
    
    @staticmethod
    def selecao(lista):
        qtd_comparacoes = 0
        qtd_troca=0
        posMenor = 0
        for i in range(len(lista)-1):
            posMenor == i
            for j in range (i+1,len(lista)): 
                qtd_comparacoes +=1
                if(lista[j]<lista[posMenor]):
                     posMenor = j
            
            if(i != posMenor):
                qtd_troca +=1
                tmp = lista[i]
                lista[i] = lista[posMenor]
                lista[posMenor] = tmp

        return qtd_comparacoes, qtd_troca

    @staticmethod
    def insercao(lista):
        qtd_comparacoes = 0
        qtd_trocas = 0

        for i in range(1, len(lista)):
            tmp = lista[i]
            j = i - 1

            while j >= 0:
                qtd_comparacoes += 1
                if tmp < lista[j]:
                    lista[j + 1] = lista[j]  # shift
                    qtd_trocas += 1
                    j -= 1
                else:
                    break

            lista[j + 1] = tmp
            # conta a "inserção final" do tmp como uma troca
            qtd_trocas += 1

        return  qtd_comparacoes, qtd_trocas

    @staticmethod
    def agitacao(lista):
        ini = 0
        fim = len(lista)
        qtd_comparacoes = 0
        qtd_trocas = 0
        while(True):
            houveTroca = False
            for i in range (ini,fim-1):
                qtd_comparacoes+=1
                if (lista[i] > lista[i+1]) :
                    qtd_trocas+=1
                    houveTroca = True
                    tmp = lista[i]
                    lista[i] = lista[i+1]
                    lista[i+1] = tmp
            if (not houveTroca ):
                break  
            fim-=1

            houveTroca = False
            for i in range (fim, ini+1,-1):
                qtd_comparacoes+=1
                if (lista[i] < lista[i-1]) :
                    qtd_trocas+=1
                    houveTroca = True
                    tmp = lista[i]
                    lista[i] = lista[i-1]
                    lista[i-1] = tmp
            ini+=1

            if (not houveTroca ):
                break         
        return qtd_comparacoes, qtd_trocas
    
    @staticmethod
    def pente(lista):
        houve_troca = True
        qtd_comparacoes = 0
        qtd_trocas = 0
        distancia = len(lista)

        while(houve_troca or distancia > 1):
            distancia = int(distancia / 1.3)
            houve_troca = False
            if distancia < 1 :
                distancia = 1
            for i in range(0,len(lista) - distancia):
                qtd_comparacoes += 1
                if(lista[i] > lista[i+distancia]):
                    qtd_trocas+=1
                    houve_troca = True
                    tmp = lista[i]
                    lista[i] = lista[i+distancia]
                    lista[i+distancia] = tmp
                           
        return qtd_comparacoes, qtd_trocas

File: Pesquisa e Ordenação/trabalho_ordenacao/TrabalhoOrdenacao/src/trabalhoordenacao/Ordenacao.java

package trabalhoordenacao;

import java.util.ArrayList;
import java.util.List;

/**
 * Classe que contém todos os métodos de ordenação
 */
public class Ordenacao {
    
    int comparacoes;
    int trocas;
    
    /**
     * Método que ordena uma lista recebida por parametro a partirdo método bolha
     * @param lista com números desordenados
     */
    public void bolha(List<Integer> lista) {
        boolean houveTroca;
        int tmp;
        int qtdComparacoes = 0, qtdTrocas = 0;
        do {
            houveTroca = false;
            for (int i = 0; i < lista.size()-1; i++){
                qtdComparacoes++;
                if (lista.get(i) > lista.get(i+1)) {
                    qtdTrocas++;
                    houveTroca = true;
                    tmp = lista.get(i);
                    lista.set(i, lista.get(i+1));
                    lista.set(i+1, tmp);
                }
            }
        } while (houveTroca);
    }
    
    /**
     * metodo que ordena uma lista por meio de seleção
     * @param lista 
     */
    public void selecao(List<Integer>lista){
        int posMenor;
        int tmp;
        int qtd_comparacoes = 0, qtd_troca=0;
        for(int i=0; i< lista.size()-1; i++){
          posMenor = i;
          for(int j = i+1; j< lista.size(); j++){
            qtd_comparacoes ++;
            if(lista.get(j)<lista.get(posMenor)){
              posMenor = j;
            }
          }
          if(i != posMenor){
            qtd_troca++;
            tmp = lista.get(i);
            lista.set(i, lista.get(posMenor));
            lista.set(posMenor,tmp);
          }  
        }
    }
    /**
     * método que ordena uma lista por meio de inserção
     */
    public void insercao(List<Integer> lista){
        int qtdComparacoes = 0, qtdTrocas = 0;
        int tmp;
        for (int i =1; i<lista.size();i++){
            tmp = lista.get(i);
            int j;
            for(j = i-1; j>= 0; j-- ){ 
                qtdComparacoes ++;
                if(tmp < lista.get(j)){
                    lista.set(j+1, lista.get(j));
                    qtdTrocas++;
                }else{
                    break;
                }
            }
            lista.set(j+1, tmp);
            qtdTrocas++;
        }
    }
    
    /**
     * método que ordena uma lista por meio de agitação
     * @param lista 
     */
    public void agitacao(List<Integer> lista) {
        boolean houveTroca = true;
        int tmp;
        int ini = 0;
        int fim = lista.size()-1;
        int qtdComparacoes = 0, qtdTrocas = 0;

        do {
            houveTroca = false;
            for (int i = 0; i < fim; i++){
                qtdComparacoes++;
                if (lista.get(i) > lista.get(i+1)) {
                    qtdTrocas++;
                    houveTroca = true;
                    tmp = lista.get(i);
                    lista.set(i,lista.get(i+1));
                    lista.set(i+1, tmp);
                }
            }
            if(!houveTroca){
              break;
            }
            fim--;

            houveTroca = false;
            for (int i = fim; i > ini+1; i--){
                qtdComparacoes++;
                if (lista.get(i) < lista.get(i-1)) {
                    qtdTrocas++;
                    houveTroca = true;
                    tmp = lista.get(i);
                    lista.set(i,lista.get(i-1));
                    lista.set(i-1,tmp);
                }
            }
            ini++;    
        } while (houveTroca);
    }
    
    /**
     * método que ordena a lista atraves do método pente
     * @param lista 
     */
    public void pente(List<Integer>lista){
        boolean houveTroca = true;
        int tmp;
        int qtdComparacoes = 0, qtdTrocas = 0;
        int distancia = lista.size();
        
        while (distancia > 1 || houveTroca ){
            distancia = (int) (distancia/1.3);
            houveTroca = false;
            if (distancia < 1){
                distancia = 1;
            }
            for (int i=0; i<lista.size()-distancia; i++){
                qtdComparacoes ++;
                if(lista.get(i) > lista.get(i+distancia)){
                    qtdTrocas ++;
                    houveTroca = true;
                    tmp = lista.get(i);
                    lista.set(i,lista.get(i+distancia));
                    lista.set(i+distancia, tmp);
                }
            }
        }
    }
    
    /**
     * método que ordena a lista com shell
     * @param lista 
     */
    public void shell(List<Integer>lista){
    int n = lista.size();
        int qtdComparacoes = 0, qtdTrocas = 0;

        int distancia = 1;
        int referenciaTamanho = 3;

        // Calcula o maior gap inicial usando a sequência de Knuth
        while (distancia < n / referenciaTamanho) {
            distancia = referenciaTamanho * distancia + 1;
        }

        while (distancia > 0) {
            // Passagem com o gap atual
            for (int i = distancia; i < n; i++) {
                int tmp = lista.get(i);
                int j;
                for (j = i - distancia; j >= 0; j -= distancia) {
                    qtdComparacoes++;
                    if (tmp < lista.get(j)) {
                        lista.set(j + distancia, lista.get(j)); // shift
                        qtdTrocas++;
                    } else {
                        break;
                    }
                }
                lista.set(j + distancia, tmp);
                qtdTrocas++;
            }
            // Reduz o gap
            distancia = distancia / referenciaTamanho;
        }
    }
    
    
    public int particionaQuick(List<Integer>lista, int ini, int fim){
        int pivo = lista.get(ini);
        int i = ini+1;
        int j = fim;
        
        while(i <= j){
            // da esquerda para direita
            while(i<=fim && lista.get(i)<=pivo){
                comparacoes++;
                i++;
            }
            //da direita para esquerda
            while(j>ini && lista.get(j)>pivo){
                comparacoes++;
                j--;
            }
            
            if (i<j){
                trocas++;
                int tmp = lista.get(i);
                lista.set(i, lista.get(j));
                lista.set(j, tmp);
            }
        }
        trocas++;
        int tmp = lista.get(ini);
        lista.set(ini, lista.get(j));
        lista.set(j, tmp);    
        return j;
    }
    
    /**
     * método que ordena uma lista utilizando quick
     * @param lista 
     */
    public void quick(List<Integer>lista, int ini, int fim){
        if (ini<fim){
            int pivo = particionaQuick(lista, ini, fim); 
            quick(lista, ini, pivo - 1);
            quick(lista, pivo + 1, fim);
        }
    }
    public static void heapSort(ArrayList<Integer> lista) {
        int tmp;
        int qtdComparacoes = 0, qtdTrocas = 0;
        int fe, fd;
        int n = lista.size();
        boolean houveTrocas;
        while (n > 1){
            houveTrocas = false;
            for (int i = 0; i*2+1 < n; i++) {
                fe = i*2+1;
                fd = i*2+2;
                qtdComparacoes++;
                
                if ( lista.get(i) < lista.get(fe) ){ //comparando o raiz com seu filho da esquerda
                    tmp = lista.get(i);
                    lista.set(i, lista.get(fe));
                    lista.set(fe, tmp);
                    qtdTrocas++;
                    houveTrocas = true;
                }
                if (fd < n) { //só vamos comparar o filho da direita se ele existir
                    qtdComparacoes++;
                    if ( lista.get(i) < lista.get(fd) ){ //comparando o raiz com seu filho da direita
                        tmp = lista.get(i);
                        lista.set(i, lista.get(fd));
                        lista.set(fd, tmp);
                        qtdTrocas++;
                        houveTrocas = true;
                    }
                }
            }
            
            if (!houveTrocas) {
                tmp = lista.get(0);
                lista.set(0, lista.get(n-1));
                lista.set(n-1, tmp);
                qtdTrocas++;
                n--;
            }
        }
        
    }
    
        
 
// ultima chave --> fecha a classe    
}
    
    
 